### PR TITLE
tests: ensure snapd can be downloaded as a module

### PR DESCRIPTION
**Pull request by atesburak**
Launchpad #1948838
File names in two folders cause an issue when trying to download snapd as a module. The files are now archived as tar files in the corresponding folder. The files don't seem be used during testing, therefore remains archived but can be re-utilized if needed.

### CODE REVIEW

**Comment by mvo5** [comment]
Thanks for working on this. If the files are truly not used we should simply remove them (git will have them in it's history). 

However looking closer at `validate-container-failures` it seems like while the test does not fail if the files are not there the files still serve a purpose (that the test does not express very well) - it checks that strange symlink loops are handled correctly by the snap validate code (i.e. snapd is not complaining about those). So I think we need to make sure that either this behavior is covered by a existing test already (so bit of digging required) or if not create a new `validate-container-happy` test that contains a simple snap that just contains the `hell` folder from this test and then some code like:
```
    snap try "$SNAP"
    snap list "$SNAP"
    ls -l "$SNAP_MOUNT_DIR/$SNAP/current/hell | MATCH "bar -> baz -> qux"
```
that checks that things are okay.

For the other strange files in the `_twisted_` dir - they are referenced in spread.ymal and commented out there because of travis issue. I think we should try to see if they work in GH actions and if so, try to re-activate them (and unpack them of course).

All of this can be done in multiple PRs of course but I think we need to investigate this a bit closer before discarding/ignoring the files.

**Comment by atesburak** [comment]
@anonymouse64 @mvo5 Can you take a look at this one and possibly trigger the actions

File: tests/completion/twisted.sh [hunk 1]
```diff
@@ -1 +1,2 @@
 cd "$SPREAD_PATH/$SPREAD_SUITE/data/twisted"
+tar -xvf $SPREAD_PATH/$SPREAD_SUITE/data/twisted/twisted.tar
```
Comment (anonymouse64) [inline tests/completion/twisted.sh:2]: newline missing from EOF

File: tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+name: test-snapd-validate-container-happy
+version: 1.0
+apps:
+  validate-container:
+    command: bin/validate-container
+plugs:
+  hell:
+    interface: content
+    target: $SNAP/hell
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml:9]: missing newline at EOF

File: tests/main/validate-container-happy/task.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+summary: check the symlinks following the right track
+
+environment:
+    SNAP: test-snapd-validate-container-happy
+
+prepare: |
+    snap pack test-snapd-validate-container-happy
+
+execute: |
+    snap try "$SNAP"
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/task.yaml:10]: hmm not sure about the order of operations here, I think that it may be better to instead have a few separate things to do and also have comments explaining what you are trying to test, I think the things that should be tested here are:

- [ ] that we can pack and install a snap that has all these weird symlinks in it (i.e. with the hell.tar extracted, then pack the snap and install the snap)
- [ ] that we can try a snap that has all these weird symlinks in it (i.e. with the hell.tar extracted, then `try` the snap)

In both cases, after the snap is installed/tried, check that the files in /snap/test-snapd-validate-container-happy/current/... have expected contents too

File: tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+name: test-snapd-validate-container-happy
+version: 1.0
+apps:
+  validate-container:
+    command: bin/validate-container
+plugs:
+  hell:
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml:7]: If the intention is to test whether these hellish symlinks are sharable/work via the content interface, then just adding these plugs here is not enough. We would need to add another snap with the hellish symlinks in them, then expose the directory with all those symlinks in them via a content interface slot and then create a second snap with a compatible content interface plug and then connect the two and ensure that the snap with the plug has the expected view

However I'm not sure the intention of testing the content interface as it pertains to these symlinks, since all the content interface is doing is bind mounting one directory over another directory and AFAIK bind mounts don't really care about how hellish symlinks are inside the directory.

If you want to exercise more things with these hellish symlinks, probably the place to look is in the layout code, where we are laying out another symlink on top of one of these hellish symlink, or perhaps laying out a hellish symlink on top of a normal symlink, but still that should work without issues since again bind mounts don't really care about the target of symlinks, just that the things being bind mounted on is of the same type as the source. 

So long story short, I would recommend against testing the content interface stuff in this PR and instead would recommend just sticking to the basics of testing `snap pack` and `snap try` with these symlinks since we do indeed have code in both of those which historically was fragile with respect to things like circular symlinks.

**Review by anonymouse64 — COMMENTED** [review]
Thanks for working on this, I think that we should reconsider the scope of this PR and just focus on fixing the immediate issue around not being able to import snapd as a module with these symlinks present, and furthermore I have some suggestions as to what to test here in the spread test since I'm unclear what exactly is currently being exercised by the spread test proposed here.

File: tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml (hunk 4 above)
Comment (atesburak) [reply tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml:7]: Makes sense to add packing the snap with the "hellish" symlinks. Switched the test flow to:

1. Pack the snap with tarred symlinks (as they cannot live in the repo)
2. "try" the snap, untar the symlinks
3. Pack the snap with the untarred symlinks
4. Install the snap and check that symlinks are still pointing to the right location

Also reverted the "twisted" test back to being disabled in the system to decouple from this PR.

File: tests/main/validate-container-happy/task.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    SNAP: test-snapd-validate-container-happy
+    SNAP_INSTALL_DIR: $(pwd)/snap-install-dir
+    SNAP_MOUNT_DIR: $(os.paths snap-mount-dir)
+
+prepare: |
+    snap pack test-snapd-validate-container-happy
+
+execute: |
+
+    # We shouldn't use relative symlinks in Github as they cannot be packed correctly.
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/task.yaml:13]: Can you unpack the symlinks before running snap try? The main code that is being exercised is the code that validates the container/skeleton of the snap is correct which only runs initially when we do snap pack and snap try, so here when you run snap try before the symlinks have been extracted nothing is being checked. 

Additionally, if you could check that the symlinks are working after running snap try that would be good to check as well, but that is not as critical as the check after we install the packed snap since snap try is really just doing a bind mount.

**Review by anonymouse64 — COMMENTED** [review]
Thanks for making the scope of this PR smaller, one more comment

File: tests/main/validate-container-happy/task.yaml (hunk 5 above)
Comment (atesburak) [reply tests/main/validate-container-happy/task.yaml:13]: Switched to this structure:

1. Untar the symlinks
2. "try" and check if symlinks exist
3. "pack" and "install" the snap
4. check to see if the symlinks are pointing to the right targets

File: tests/main/validate-container-happy/task.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    # Untar the symlinks 
+    tar -xvf "$SNAP"/hell/hell.tar -C "$SNAP"/hell
+
+    snap try "$SNAP"
+    # Check to see if the symlinks exist
+    ls -l "$SNAP"/hell | MATCH "foo -> baz -> qux"
+    ls -l "$SNAP"/hell | MATCH "foo -> bar -> qux"
+    ls -l "$SNAP"/hell | MATCH "bar -> baz -> qux"
+    ls -l "$SNAP"/hell | MATCH "foo -> bar -> baz"
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/task.yaml:27]: these checks should be for $SNAP_MOUNT_DIR 

```suggestion
    ls -l "$SNAP_MOUNT_DIR"/hell | MATCH "foo -> baz -> qux"
    ls -l "$SNAP_MOUNT_DIR"/hell | MATCH "foo -> bar -> qux"
    ls -l "$SNAP_MOUNT_DIR"/hell | MATCH "bar -> baz -> qux"
    ls -l "$SNAP_MOUNT_DIR"/hell | MATCH "foo -> bar -> baz"
```

File: tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml (hunk 4 above)
Comment (anonymouse64) [reply tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml:7]: thanks, can you remove the `plugs` section from this snap.yaml too?

File: tests/main/validate-container-happy/task.yaml (hunk 6 above)
Comment (anonymouse64) [reply tests/main/validate-container-happy/task.yaml:27]: also I have a slight preference for checking the symlinks with readlink instead of `ls -l`

**Review by anonymouse64 — COMMENTED** [review]
a few more little nitpicks

File: tests/main/validate-container-happy/task.yaml (hunk 6 above)
Comment (atesburak) [reply tests/main/validate-container-happy/task.yaml]: I prefer readlink myself too, just wanted to diversify the checks. Switched the checks during the "try" to readlink as well.

File: tests/main/validate-container-happy/task.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    # Check to see if the symlinks exist
+    readlink "$SNAP_MOUNT_DIR"/"$SNAP"/hell/bar | MATCH "foo -> baz -> qux"
+    readlink "$SNAP_MOUNT_DIR"/"$SNAP"/hell/baz | MATCH "foo -> bar -> qux"
+    readlink "$SNAP_MOUNT_DIR"/"$SNAP"/hell/foo | MATCH "bar -> baz -> qux"
+    readlink "$SNAP_MOUNT_DIR"/"$SNAP"/hell/qux | MATCH "foo -> bar -> baz"
+
+    # Create a new snap structure that includes the unpacked symlinks
+    mkdir -p "$SNAP_INSTALL_DIR"
+    cp -ra "$SNAP"/* "$SNAP_INSTALL_DIR"
+    cp -ra "$SNAP_MOUNT_DIR"/"$SNAP"/hell "$SNAP_INSTALL_DIR"
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/task.yaml:33]: I suppose this is fine since you are using `cp -a` and I trust that `cp` is doing the right thing but for clarity, this could section could just be:

```
snap pack "$SNAP"
```

File: tests/main/validate-container-happy/task.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    snap pack "$SNAP_INSTALL_DIR"
+    snap install --dangerous test-snapd-validate-container-happy_1.0_all.snap
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/task.yaml:36]: Ideally we would remove the snap that we installed in try mode with `snap try` before installing this .snap file, since both are for the same snap and we essentially end up doing a refresh which is more complicated of a test than we need for here, so my preference would be to do:

1. unpack tarball
2. snap try the dir
3. check symlinks exist in /snap/...
4. remove snap
5. snap pack the dir
6. snap install the .snap file
7. check symlinks exist in /snap/...

does that make sense?

Comment (atesburak) [reply tests/main/validate-container-happy/task.yaml:32]: Simplified the process accordingly. Current steps:

1. Untar the symlinks to the existing folder
2. "try" the folder as a new snap
3. Check that symlinks point to the right target
4. Remove the snap
5. Pack the directory as a new snap
6. Install the snap
7. Check that symlinks point to the right target
8. Remove the snap

File: tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+name: test-snapd-validate-container-happy
+version: 1.0
+apps:
+  validate-container:
+    command: bin/validate-container
+
```
Comment (bboozzoo) [inline tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml:6]:
```suggestion
```

File: tests/main/validate-container-happy/task.yaml [hunk 10]
```diff
@@ -0,0 +1,37 @@
+summary: check the symlinks following the right track
+
+environment:
+    SNAP: test-snapd-validate-container-happy
+    SNAP_MOUNT_DIR: $(os.paths snap-mount-dir)
```
Comment (bboozzoo) [inline tests/main/validate-container-happy/task.yaml:5]: Hm isn't this evaluated on the host? So if I use Fedora locally, but try to run the spread test on ubuntu-20.04-64, it will fail because a different mount dir will be expected?

File: tests/main/validate-container-happy/task.yaml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+prepare: |
+
+execute: |
+
```
Comment (bboozzoo) [inline tests/main/validate-container-happy/task.yaml:10]:
```suggestion
    SNAP_MOUNT_DIR="$(os.paths snap-mount-dir)"
```

File: tests/main/validate-container-happy/task.yaml (hunk 10 above)
Comment (bboozzoo) [inline tests/main/validate-container-happy/task.yaml:5]:
```suggestion
```

File: tests/main/validate-container-happy/task.yaml [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    # Untar the symlinks 
+    tar -xvf "$SNAP"/hell/hell.tar -C "$SNAP"/hell
+
+    snap try "$SNAP"
+    # Check to see if the symlinks point to the right paths
+    readlink "$SNAP"/hell/bar | MATCH "foo -> baz -> qux"
+    readlink "$SNAP"/hell/baz | MATCH "foo -> bar -> qux"
+    readlink "$SNAP"/hell/foo | MATCH "bar -> baz -> qux"
+    readlink "$SNAP"/hell/qux | MATCH "foo -> bar -> baz"
```
Comment (bboozzoo) [inline tests/main/validate-container-happy/task.yaml:25]: Did you mean `"$SNAP_MOUNT_DIR"/"$SNAP"/current` ?

File: tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+name: test-snapd-validate-container-happy
+version: 1.0
+apps:
+  validate-container:
+    command: bin/validate-container
```
Comment (bboozzoo) [inline tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml:5]: Hmm funny, looks like it's missing final `\n`, so if you were to `cat` the file in the terminal you'd end up with your prompt following the content of the last time.

Comment (atesburak) [reply tests/main/validate-container-happy/test-snapd-validate-container-happy/meta/snap.yaml]: Fixed it, thanks :)

**Review by bboozzoo — APPROVED** [review]
LGTM. A possible followup to consider could drop the tar completely and instead create the desired structure directly in the spread test, but what is here is sufficient to land and unblock ubuntu-image.

File: tests/main/validate-container-happy/task.yaml [hunk 14]
```diff
@@ -0,0 +1,38 @@
+summary: check the symlinks following the right track
+
+environment:
+    SNAP: test-snapd-validate-container-happy
+
+prepare: |
+
```
Comment (anonymouse64) [inline tests/main/validate-container-happy/task.yaml:7]: in a followup please delete this section since there is no code being executed in it

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for iterating on this
